### ISSUE TITLE
RFE: configure, execution, and data separation needed

### ISSUE THREAD

**Issue by taw00** [nmarley/sentinel#21]
Everything resides in the solitary sentinel tree and is largely inseparable. For testing, it's fine, but for the long haul, adherence to best practice, or at least better practice needs to occur.

* Config: sentinel.conf
* Data: database/
* Execution: most of the rest

And of course, ideally one can specify the various targets from the commandline, either with options or environment variables (though the environment variable route is rather ugly and frowned upon).
crontab.py --dashcore-conf=/path/to/dash.conf --sentinel-conf=/path/to/sentinel.conf --db=/path/to/variable/data

And why named crontab.py? Why not... "sentinel" or "dashcore-sentinel" or some such?

Anyway...
So, if you want to pile up everything, by default, into a solitary directory... sure... but these elements need to be separable. For one, that's simply best and standard practice, and secondarily you can do interesting things then, like have one installation of the code... serving several masternodes, etc.

If I find time, I may dust of my python and tackle this myself, but it would be awhile, so filing this RFE here.

Thanks! -t0dd

**Comment by nmarley** [comment]
Thanks for submitting.

Can you be more specific in how you'd like to see it structured? Can you reference these "best/better" practices? I'm all for improving, but don't know which practices you're referring to currently.

Database info currently stored in `database`, config is only one file, which sits at the root, and since the project isn't planned to grow into a huge monolith(er), I don't think having a separate `config` directory for a single config would help anything. Just noticed that `config.py` sits at the root also, but that's mostly left over from almost a year ago and the actual config info was broken out into `sentinel.conf`.

It's pretty standard for LICENSE and README files to sit at the root as they are now, as well as the `requirements.txt` Python manifest. Tests are in their own directory, which I believe is also a pretty standard thing to do (although I admit they need some work).

> And why named crontab.py? Why not... "sentinel" or "dashcore-sentinel" or some such?

The original author called it that waay back in this commit: 63514cbc4218268897a9be37cace9839d1ef0ff5. I haven't really seen anything wrong with it, so haven't had any reason to change it.

> So, if you want to pile up everything, by default, into a solitary directory... sure... but these elements need to be separable. For one, that's simply best and standard practice, and secondarily you can do interesting things then, like have one installation of the code... serving several masternodes, etc.

Just addressed most of this, but basically it's not all piled into the root -- there are several directories for the different functionality, `lib`, `test`, `scripts`, `database`.

As far as one installation serving several masternodes, it's not really designed to work that way, but I guess it could as long as you could specify a path to a different config file, using a different database, in the crontab script.

In summary, if you want to open a separate issue for specifying config paths, feel free to do so. Mostly this was focused on the directory/file structure, so please let me know specifically what you don't think matches any standard or best practice, as I'm curious to find out what you're seeing that I'm not.

**Comment by taw00** [comment]
I will try to put response together that is more comprehensive another day when I have more time, but just about any piece of software should be adaptable to things like the File System Hierarchy Standard https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard

That particular standard is linux focused but each OS follows the spirit of the standard. I.e., Dependent on OS packaging, software is able to split its components into executable bits, library bits, configuration, and then variable bits (like the database). On Linux OSes (and Unix'es and BSDes like Apple), if one were to lock down the system most securely, the more static elements, for example likely reside on a read only part of the OS, the database would sit in another partition that has different mount rules, etc. LICENSE and doc files get stored in standard docs directory structures... but that is trivial to do now since nothing depends on them.

Anyway. The shipped tarball and builds can make assumptions about "one tree" if desired, but the configuration should allow for these splits to be recognized. It could be as simple as adding target paths to the conf file and then it is up to those who deploy it, or 3rd party packagers to shuffle the defaults to whatever the demands of the particular OS are and allow things like 1 install to serve many configurations.

A simple example is ... a mysql install on an OS is _one_ installation for the application, but can have infinite database instances for different users. Or in the case of a user application, you can have 5 different users of an OS (your family on your household desktop for example), but there is only one installation of Firefox, not 5. Etc. etc. And with Sentinel, if I wanted to deploy 20 Masternodes on a beefy box... one currently would have to deploy 20 Sentinel installs which is not "correct" versus 1 install that serves 20 masternodes leveraging 20 Sentinel database instantiations. Note, as another example datapoint, dash nodes correctly only need 1 software installation, no matter how many instantiations I create on that box. 

Anyway. When I have more time, I can try to put together more expansive thoughts on how it can be better structured and/or addressed in configuration. Again. Not a "must do it today" item, but more of a polish element to how Sentinel should be deployable over time especially as it slowly increases in function and complexity. It will also help govern design more appropriately though right now you have the elements of separation already in place, just not that final configurability.

Note: I am [building Sentinel now as a RPM package for Red Hat Linux -family of OSes](https://github.com/taw00/dashcore-rpm). And in the process tripped across this since, of course, like all Linuxes they follow the standard mentioned at the top of this comment. The FSHS is simply Linux'es spin on the similar standards that all Oses follow when they package software to their specific packaging systems.

Thanks for your response and I hope this helps address some of your questions. And, like I said, I may carve out some time to submit the changes myself, because for the simplest separation of function... it would not be complex to implement.

**Comment by nmarley** [comment]
Ah, I see, this is about the RPM packages... I understand the FHS  well (and criticisms against it). For systems applications and daemons the FHS-style distribution is normal, but have you ever seen a web app, say, for example, a Rails app deployed in this manner?

I personally think that having to manage packages spread out across the filesystem is much more difficult than having the entire app isolated to one area. The current method of upgrading Sentinel is simply a `git pull`, and the current method to downgrade is equally simple: `git checkout <commit/tag>`. Since everything's all together, if a user wants to remove Sentinel completely & start over, they just remove the directory and re-clone.

> And with Sentinel, if I wanted to deploy 20 Masternodes on a beefy box... one currently would have to deploy 20 Sentinel installs which is not "correct" versus 1 install that serves 20 masternodes leveraging 20 Sentinel database instantiations. Note, as another example datapoint, dash nodes correctly only need 1 software installation, no matter how many instantiations I create on that box.

This is a valid point (and I believe a separate one from the FHS) and I believe can be implemented as an option to the main script, to point to an alternative `sentinel.conf`. It should be noted that each Sentinel installation must have it's own separate database.

The biggest problem in your hypothetical example would probably be finding 20 IPv4 addresses to allocate to each masternode.

**Comment by taw00** [comment]
I understand your point, but the standard is the standard. I'm not asking you to implement it. I am asking you to enable it for packagers. a --conf and a --db would be fine for now.

As for webapps. Yes. In fact, I have worked on several significantly large web-apps that were packaged, digitally signed, and deployed in such a manner. Python and Ruby on Rails, in fact.

But you are correct, many people who build their own webapps, don't bother. But folks that build "product" that ships to customers... generally do. Now, one could argue that the Dash project is just a project that doesn't need that level of polish, but... well... We want to be flexible. We want to eventually get to the point were you deploy the code native to your package manager. RPM is just one example. There will be work to container-ize Dash as well.

Anyway. It's what is expected. And if I get some time to carve off, I will just implement it myself, but that won't be for a bit. No rush. But it is a need, IMHO. And doesn't take much to make it serviceably meet the standards. Thanks.

**Comment by nmarley** [comment]
> I understand your point, but the standard is the standard. I'm not asking you to implement it. I am asking you to enable it for packagers.

It is "a standard" and not "the *only* standard", but I don't want to get into a war of philosophy or religion here. ;)

> a --conf and a --db would be fine for now.

Actually just a --conf should work, as the DB is specified in the sentinel.conf file.

I'm trying to get some of these implemented this weekend (e.g. `bin/sentinel.py` instead of `scripts/crontab.py`, since the current name doesn't make sense). I'll try and get to the `--conf` option if at all possible and that should enable you (or anyone else) to build a .rpm package. I believe this is all that should be needed in order to enable this.

**Comment by taw00** [comment]
Sounds good. Thanks for listening. In the end, this is a current "nice to have" but if there is a nice time to have... it's now. ;)

Anyway. Again, thanks for consideration.

**Comment by nmarley** [comment]
As of commit f931195a222c1484a18f0bcae4dd6f49dd3b9f50, it is now possible to specify an alternative `sentinel.conf` file by setting the `SENTINEL_CONFIG` environment variable to the alternative path.
